### PR TITLE
Refactoring reporting ID capability in patterned mesh generators to support multiple ID assignment

### LINKED ISSUE #26858: Refactoring reporting ID capability in patterned mesh generators to support multiple ID assignments
## Reason

The current reporting ID capability in patterned mesh generators (`PatternedCartesianMeshGenerator` and `PatternedHexMeshGenerator`) only allows to assign a single reporting ID.

This update will support assigning multiple reporting IDs in a patterned mesh generation of the reactor module. For example, in the reactor mesh generation, users can assign two reporting IDs for individual pin cells and for individual pin types in an assembly. Additionally, users can assign multiple manual reporting IDs to patterned meshes.

## Design

The input structures for assigning multiple reporting ID will be:
```
  type = PatternedHexMeshGenerator
  id_name = 'cell_id, pattern_id, manual_1_id, manual_2_id'
  assign_type 'cell patten manual manual'
  id_pattern = '1 1;
               2 2 2;
                3 3|
                1 2;
               1 2 3
                2 3;
```

## Impact

The patterned mesh generators will support multiple reporting ID assignments. This will also affect the RGMB implementation - current implementation can be simplified by utilizing this update.

### PR DESCRIPTION
**Pull request by yjung-anl**
closes #26858

### CODE REVIEW

**Comment by yjung-anl** [comment]
Now, this PR is ready for review.

File: modules/reactor/doc/content/source/meshgenerators/PatternedCartesianMeshGenerator.md [hunk 1]
```diff
@@ -71,6 +71,23 @@ The ID values themselves are stored as extra element integers on the mesh.
 For example, the reporting IDs for individual pins (`pin_id`) can be assigned when assemblies are built because the IDs for pin level are uniquely determined from the pin arrangement within each assembly type.
 Similarly, the assembly reporting IDs (`assembly_id`) are assigned in the core construction process.
 
+The multiple `reporting IDs` can be assigned by defining the multiple name of the reporting ID variable, which are provided through the[!param](/Mesh/PatternedCartesianMeshGenerator/id_name).
```
Comment (GiudGiud) [inline modules/reactor/doc/content/source/meshgenerators/PatternedCartesianMeshGenerator.md:74]:
```suggestion
The multiple `reporting IDs` can be assigned by defining the multiple names of the reporting ID variable, which are provided through the[!param](/Mesh/PatternedCartesianMeshGenerator/id_name).
```
it s not a variable though. do you mean parameter here or EEID?

File: modules/reactor/doc/content/source/meshgenerators/PatternedCartesianMeshGenerator.md [hunk 2]
```diff
@@ -71,6 +71,23 @@ The ID values themselves are stored as extra element integers on the mesh.
...
+The correponding assignment scheme should be provided in [!param](/Mesh/PatternedCartesianMeshGenerator/assign_type) for each reporting ID names, accordingly.
```
Comment (GiudGiud) [inline modules/reactor/doc/content/source/meshgenerators/PatternedCartesianMeshGenerator.md:75]:
```suggestion
The corresponding assignment scheme should be provided in [!param](/Mesh/PatternedCartesianMeshGenerator/assign_type) for each reporting ID names, accordingly.
```

File: modules/reactor/doc/content/source/meshgenerators/PatternedCartesianMeshGenerator.md [hunk 3]
```diff
@@ -71,6 +71,23 @@ The ID values themselves are stored as extra element integers on the mesh.
...
+In case of multiple `manual` assign types are used, the same number of manual ID patterns should be provided in [!param](/Mesh/PatternedCartesianMeshGenerator/id_pattern).
```
Comment (GiudGiud) [inline modules/reactor/doc/content/source/meshgenerators/PatternedCartesianMeshGenerator.md:76]:
```suggestion
In the case that multiple `manual` `assign_type`s are used, the same number of manual ID patterns should be provided in [!param](/Mesh/PatternedCartesianMeshGenerator/id_pattern).
```

File: modules/reactor/doc/content/source/meshgenerators/PatternedCartesianMeshGenerator.md [hunk 4]
```diff
@@ -71,6 +71,23 @@ The ID values themselves are stored as extra element integers on the mesh.
...
+Each manual pattern in [!param](/Mesh/PatternedCartesianMeshGenerator/id_pattern) should be separated by using the delimiter '|'.
+These defined ID patterns are sequentially assigned to the reporting IDs having `manual` assignment scheme.
+The below is an example of using multiple `reporting ID` assignment.
+Here, `manual_1_id` uses the first pattern in defined in [!param](/Mesh/PatternedCartesianMeshGenerator/id_pattern), and `manaul_2_id` uses the second one.
```
Comment (GiudGiud) [inline modules/reactor/doc/content/source/meshgenerators/PatternedCartesianMeshGenerator.md:80]:
```suggestion
Here, `manual_1_id` uses the first pattern in defined in [!param](/Mesh/PatternedCartesianMeshGenerator/id_pattern), and `manual_2_id` uses the second one.
```

File: modules/reactor/doc/content/source/meshgenerators/PatternedCartesianMeshGenerator.md [hunk 5]
```diff
@@ -71,6 +71,23 @@ The ID values themselves are stored as extra element integers on the mesh.
...
+
+!listing!
+id_name = 'manual_1_id cell_id manaul_2_id'
+assign_type 'manual cell manual'
+id_pattern = '1 2;
+              3 4|
+              4 3;
+              2 1;
+!listing-end!
+
```
Comment (GiudGiud) [inline modules/reactor/doc/content/source/meshgenerators/PatternedCartesianMeshGenerator.md:90]: apply comments to the next file

File: modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests [hunk 6]
```diff
@@ -79,5 +79,12 @@
       detail = 'for Cartesian cores with regular square core boundary'
       recover = false
     []
+    [multi_id]
+      type = 'Exodiff'
+      input = 'patterned_cartesian_assembly_reporting_multi_id.i'
+      exodiff = 'patterned_cartesian_assembly_reporting_multi_id_out.e'
+      detail = 'for multiple differnt assignmet schemes'
```
Comment (GiudGiud) [inline modules/reactor/test/tests/meshgenerators/reporting_id/cartesian_id/tests:86]:
```suggestion
      detail = 'for having different reporting ID assignment schemes for each pattern'
```

File: modules/reactor/src/meshgenerators/PatternedCartesianMeshGenerator.C [hunk 7]
```diff
@@ -104,14 +104,16 @@ PatternedCartesianMeshGenerator::validParams()
...
-  params.addParam<MooseEnum>("assign_type", option, "Type of integer ID assignment");
-  params.addParam<std::vector<std::vector<dof_id_type>>>(
+  std::vector<MooseEnum> option = {MooseEnum("cell pattern manual", "cell")};
+  params.addParam<std::vector<MooseEnum>>(
+      "assign_type", option, "List of integer ID assignment types");
+  params.addParam<std::vector<std::vector<std::vector<dof_id_type>>>>(
       "id_pattern",
-      "User-defined element IDs. A double-indexed array starting with the upper-left corner");
+      "User-defined element IDs. A double-indexed array starting with the upper-left corner. In "
+      "case of providing multiple pattern, each pattern should be separated using '|'");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PatternedCartesianMeshGenerator.C:116]:
```suggestion
      "User-defined element IDs. A double-indexed array starting with the upper-left corner. When "
      "providing multiple patterns, each pattern should be separated using '|'");
```

File: modules/reactor/include/meshgenerators/PatternedCartesianMeshGenerator.h [hunk 8]
```diff
@@ -78,14 +78,16 @@ class PatternedCartesianMeshGenerator : public PolygonMeshGeneratorBase
   std::vector<SubdomainName> _peripheral_block_names;
   /// Whether reporting ID is added to mesh
   const bool _use_reporting_id;
+  /// names of reporting ID
+  std::vector<std::string> _reporting_id_name;
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PatternedCartesianMeshGenerator.h:82]:
```suggestion
  std::vector<std::string> _reporting_id_names;
```

File: modules/reactor/include/meshgenerators/PatternedCartesianMeshGenerator.h [hunk 9]
```diff
@@ -78,14 +78,16 @@ class PatternedCartesianMeshGenerator : public PolygonMeshGeneratorBase
...
   /// reporting ID assignment type
-  const ReportingIDGeneratorUtils::AssignType _assign_type;
+  std::vector<ReportingIDGeneratorUtils::AssignType> _assign_type;
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PatternedCartesianMeshGenerator.h:84]:
```suggestion
  std::vector<ReportingIDGeneratorUtils::AssignType> _assign_types;
```

File: modules/reactor/include/meshgenerators/PatternedCartesianMeshGenerator.h [hunk 10]
```diff
@@ -78,14 +78,16 @@ class PatternedCartesianMeshGenerator : public PolygonMeshGeneratorBase
...
   /// flag to indicate if exclude_id is defined
   const bool _use_exclude_id;
   /// vector indicating which ids in the pattern to exclude (true at pattern positions to exclude)
   std::vector<bool> _exclude_ids;
-  /// hold reporting ID for each input pattern cell
-  std::vector<std::vector<dof_id_type>> _id_pattern;
+  /// hold ID patterns for each manual reporting ID. Individual ID pattern contains ID values for each pattern cell.
+  std::map<std::string, std::vector<std::vector<dof_id_type>>> _id_pattern;
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PatternedCartesianMeshGenerator.h:90]:
```suggestion
  std::map<std::string, std::vector<std::vector<dof_id_type>>> _id_patterns;
```

File: modules/reactor/src/meshgenerators/PatternedCartesianMeshGenerator.C [hunk 11]
```diff
@@ -952,29 +974,35 @@ PatternedCartesianMeshGenerator::addReportingIDs(
...
+    unsigned int extra_id_index;
+    if (!mesh->has_elem_integer(element_id_name))
+      extra_id_index = mesh->add_elem_integer(element_id_name);
+    else
+    {
+      extra_id_index = mesh->get_elem_integer_index(element_id_name);
+      paramWarning(
+          "id_name", "An element integer with the name '", element_id_name, "' already exists");
+    }
+    // assign reporting IDs to individual elements
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PatternedCartesianMeshGenerator.C:990]:
```suggestion
    }
    
    // assign reporting IDs to individual elements
```

File: modules/reactor/src/meshgenerators/PatternedCartesianMeshGenerator.C [hunk 12]
```diff
@@ -254,14 +254,36 @@ PatternedCartesianMeshGenerator::PatternedCartesianMeshGenerator(const InputPara
...
+    // get reporting id name input
+    _reporting_id_name = getParam<std::vector<std::string>>("id_name");
+    const unsigned int num_reporting_ids = _reporting_id_name.size();
+    // get reporting id assign type input
+    const auto input_assign_type = getParam<std::vector<MooseEnum>>("assign_type");
+    if (input_assign_type.size() != num_reporting_ids)
+      paramError("assign_type", "This parameter must have a length equal to length of id_name.");
+    // list of reporting id names using manual id patterns;
+    std::vector<std::string> manual_ids;
+    for (unsigned int i = 0; i < num_reporting_ids; ++i)
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PatternedCartesianMeshGenerator.C:266]:
```suggestion
    for (const auto i : make_range(num_reporting_ids))
```

File: modules/reactor/src/meshgenerators/PatternedHexMeshGenerator.C [hunk 13]
```diff
@@ -106,14 +106,16 @@ PatternedHexMeshGenerator::validParams()
   params.addParam<bool>("deform_non_circular_region",
                         true,
                         "Whether the non-circular region (outside the rings) can be deformed.");
-  params.addParam<std::string>("id_name", "Name of extra integer ID set");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PatternedHexMeshGenerator.C:109]: same as other file

**Review by GiudGiud — CHANGES_REQUESTED** [review]
looks good

**Comment by yjung-anl** [comment]
@GiudGiud  Thank you!!

**Comment by yjung-anl** [comment]
@GiudGiud All comments were addressed.

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**